nep-22.mediawiki

```diff
@@ -0,0 +1,123 @@
+
+==Abstract==
+
+This proposal outlines a basic method guideline for the NEO blockchain that will provide systems with a generalized interaction mechanism for smart contract initial deploy, update and destroy.
+
+==Motivation==
+
+As the NEO blockchain scales, Smart Contract Initial Deploy, Update and Destroy will become increasingly important.  Without guideline for <code>update</code> and <code>destroy</code> method, systems will be required to maintain a unique API for each contract, regardless of their similarity to other contracts which makes application development very inconvenient. This guideline can lead most contracts to implement the same basic methods with the same types of parameters but still accept diversity.
+
+==Specification==
+
+In the method definitions below, we provide both the definitions of the functions as they are defined in the contract as well as the invoke parameters.
+
+===Methods===
+
+====_initialize====
+
+It's an internal function which will be compiled into Abi file if any function is using initialize statement. These functions which have initialize statement will be executed first when loading contract. The format should be [InitialValue(Value, ContractParameterType)], here's an example.
+
+<pre>
+[InitialValue("NdUL5oDPD159KeFpD5A9zw5xNF1xLX6nLT", ContractParameterType.Hash160)]
+private static readonly UInt160 owner = default;
+</pre>
+
+The <code>owner</code> here will be initialized as <code>"NdUL5oDPD159KeFpD5A9zw5xNF1xLX6nLT"</code> everytime when the contract be executed.
+
+====_deploy====
+
+<pre>
+{
+  "name": "_deploy",
+  "safe": false,
+  "parameters": [
+    {
+      "name": "data",
+      "type": "Any"
+    },
+    {
+      "name": "update",
+      "type": "Boolean"
+    }
+  ],
+  "returntype": "Void"
+}
+</pre>
+
+This is an optional function which will be automatically executed by ContractManagement contract when a contract is first deployed or updated.
+
+The parameter <code>data</code> can be any type of supported parameters for contract-specific purpose. It's the same <code>data</code> that is accepted by Deploy() and Update() in ContractManagement contract.
+
+<code>update</code> is true when contract is updated via <code>update()</code> and it's false on initial deployment via <code>_deploy()</code>.
+
+====verify====
+
+<pre>
+{
+  "name": "verify",
+  "safe": false,
+  "parameters": [],
+  "returntype": "bool"
+}
+</pre>
+
+If this contract address is included in the transaction signature, it MUST apply this. This method will be triggered as a VerificationTrigger to verify that the signature is correct.. Here's an example:
+
+<code>public static bool Verify() => Runtime.CheckWitness(Owner);</code>
+
+This method needs to be called when withdrawing token from the contract.
+
+====update====
+
+<pre>
+{
+  "name": "update",
+  "safe": false,
+  "parameters": [
+    {
+      "name": "nefFile",
+      "type": "ByteArray"
+    },
+    {
+      "name": "manifest",
+      "type": "String"
```

Comment:
How come it's a `String` btw? ContractManagement itself expects a `ByteArray`.

Reply:
It doesn't matter for arguments. In the vm its already byte array; to be specific `ByteString`; can be anything for the most part. It all get compiled down to byte array.

Reply:
That's not the point, these are different things as per NEP-14 and one can expect some type symmetry between `ContractManagement` definition of `update` and this one. Manifests matter, especially if you try to generate some code using them.

Reply:
Then why not make a `ByteString`?

Reply:
> That's not the point, these are different things as per NEP-14 and one can expect some type symmetry between `ContractManagement` definition of `update` and this one. Manifests matter, especially if you try to generate some code using them.

Fixed